projects/lesson-3/runtime/src/kitties.rs

```diff
@@ -1,37 +1,197 @@
-use support::{decl_module, decl_storage, StorageValue, StorageMap};
+use support::{decl_module, decl_storage, StorageValue, StorageMap, ensure, dispatch::Result};
 use codec::{Encode, Decode};
 use runtime_io::blake2_128;
 use system::ensure_signed;
+use sr_primitives::traits::Hash;
+use rstd::cmp;
 
-pub trait Trait: system::Trait {
+pub trait Trait: system::Trait + balances::Trait {
 }
 
-#[derive(Encode, Decode, Default)]
-pub struct Kitty(pub [u8; 16]);
+// #[derive(Encode, Decode, Default)]
+// pub struct Kitty(pub [u8; 16]);
+
+#[derive(Debug, Encode, Decode, Default, Clone, PartialEq)]
+pub struct Kitty<Hash, Balance> {
+    id : Hash,
+	dna : [u8; 16],
+	price : Balance,
+	gen : u64,
+}
 
 decl_storage! {
 	trait Store for Module<T: Trait> as Kitties {
 		/// Stores all the kitties, key is the kitty id / index
-		pub Kitties get(kitty): map u32 => Kitty;
+		//pub Kitties get(kitty): map u32 => Kitty;
 		/// Stores the total number of kitties. i.e. the next kitty index
-		pub KittiesCount get(kitties_count): u32;
+		//pub KittiesCount get(kitties_count): u32;
+
+		Kitties get(kitty): map T::Hash => Kitty<T::Hash, T::Balance>;
+		KittyOwner get(owner_of): map T::Hash => Option<T::AccountId>;
+
+		AllKittiesArray get(kitty_by_index): map u64 => T::Hash;
+		AllKittiesCount get(all_kitties_count): u64;
+		AllKittiesIndex: map T::Hash => u64;
+
+		OwnedKittiesArray get(kitty_of_owner_by_index): map (T::AccountId, u64) => T::Hash;
+		OwnedKittiesCount get(owned_kitty_count): map T::AccountId => u64;
+		OwnedKittiesIndex: map T::Hash => u64;
+
+		Nonce: u128;
+
+		
 	}
 }
 
 decl_module! {
 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
 		/// Create a new kitty
-		pub fn create(origin) {
-			let sender = ensure_signed(origin)?;
-			let count = Self::kitties_count();
-			if count == u32::max_value() {
+		// pub fn create(origin) {
+		// 	let sender = ensure_signed(origin)?;
+		// 	let count = Self::kitties_count();
+		// 	if count == u32::max_value() {
+		// 		return Err("Kitties count overflow");
+		// 	}
+		// 	let payload = (<system::Module<T>>::random_seed(), sender, <system::Module<T>>::extrinsic_index(), <system::Module<T>>::block_number());
+		// 	let dna = payload.using_encoded(blake2_128);
+		// 	let kitty = Kitty(dna);
+		// 	Kitties::insert(count, kitty);
+		// 	KittiesCount::put(count + 1);
+		// }
+
+		// Initializing events
+		// this is needed only if you are using events in your module
+	    //fn deposit_event() = default;
+
+		fn create_kitty(origin) -> Result {
+		    let sender = ensure_signed(origin)?;
+
+			let count = Self::all_kitties_count();
+			if count == u64::max_value() {
```

Comment:
No need for this one, since you already checked it in following `all_kitties_count.checked_add(1)`